File: 07-Encordados-JPA-Hibernate-Java/src/main/java/models/Raqueta.java
package models;

import jakarta.persistence.*;
import org.hibernate.annotations.UuidGenerator;

import java.util.Objects;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "raquetas")
@NamedQuery(name = "Raqueta.findAll", query = "SELECT r FROM Raqueta r")
public class Raqueta {
    private UUID uuid = UUID.randomUUID();
    private String marca;
    private Double precio;
    // Mi lista de tenistas
    private Set<Tenista> tenistas;

    public Raqueta(String marca, Double precio) {
        this.marca = marca;
        this.precio = precio;
    }

    public Raqueta() {
    }

    @Id
    @GeneratedValue
    @UuidGenerator // Genera un UUID
    @Column(name = "uuid")
    public UUID getUuid() {
        return uuid;
    }

    public void setUuid(UUID uuid) {
        this.uuid = uuid;
    }

    public String getMarca() {
        return marca;
    }

    public void setMarca(String marca) {
        this.marca = marca;
    }

    public Double getPrecio() {
        return precio;
    }

    public void setPrecio(Double precio) {
        this.precio = precio;
    }

    // 1-N con Tenista, una raqueta puede tener muchos tenistas que la usan por patrocinio
    // No queremos que se borre en cascada porque no queremos que se borren los tenistas
    // @OneToMany(orphanRemoval = true, cascade = CascadeType.PERSIST, mappedBy = "orderRequest")
    @OneToMany(mappedBy = "raqueta", orphanRemoval = true,  cascade = CascadeType.PERSIST)
    public Set<Tenista> getTenistas() {
        return tenistas;
    }

    public void setTenistas(Set<Tenista> tenistas) {
        this.tenistas = tenistas;
    }

    // Metodo helpers
    public void addTenista(Tenista tenista) {
        //tenistas.add(tenista);
        tenista.setRaqueta(this);
    }

    public void removeTenista(Tenista tenista) {
        tenista.removeRaqueta();
    }

    @Override
    public String toString() {
        return "Raqueta{" +
                "uuid=" + uuid +
                ", marca='" + marca + '\'' +
                ", precio=" + precio +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Raqueta)) return false;
        Raqueta raqueta = (Raqueta) o;
        return uuid.equals(raqueta.uuid) && marca.equals(raqueta.marca) && precio.equals(raqueta.precio);
    }

    @Override
    public int hashCode() {
        return Objects.hash(uuid, marca, precio);
    }
}

File: 08-Encordados-JPA-Hibernate-Kotlin/src/main/kotlin/models/Raqueta.kt
package models

import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.GenericGenerator
import org.hibernate.annotations.Type
import java.time.LocalDateTime
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "raquetas")
// Defino estas queries que estarán disponibles globalmente
@NamedQuery(name = "Raqueta.findAll", query = "SELECT r FROM Raqueta r")
data class Raqueta(
    @Id @GeneratedValue
    // @UuidGenerator // Hibernate 6
    // Hibernate 5
    @GenericGenerator(
        name = "UUID",
        strategy = "org.hibernate.id.UUIDGenerator",
    )
    @Column(name = "uuid")
    @Type(type = "uuid-char")
    val uuid: UUID = UUID.randomUUID(),
    var marca: String,
    var precio: Double,

    // Opcional, solo si queremos bidireccionalidad
    // 1-N con Tenista, una raqueta puede tener muchos tenistas que la usan por patrocinio
    // No queremos que se borre en cascada porque no queremos que se borren los tenistas
    // No pengais persistir en cascada porque no queremos que se persistan los tenistas
    // @OneToMany(orphanRemoval = true, cascade = CascadeType.PERSIST, mappedBy = "orderRequest")
    @OneToMany(mappedBy = "raqueta", orphanRemoval = true, fetch = FetchType.EAGER)
    val tenistas: MutableList<Tenista> = mutableListOf(),

    // Mi representante
    @Embedded
    var representante: Representante? = null,

    @Column(name = "created_at")
    @Type(type = "org.hibernate.type.LocalDateTimeType")
    @CreationTimestamp
    val createdAt: LocalDateTime = LocalDateTime.now(),

    )

// No voy a guardar para las raquetas los tenistas, porque eso lo puedo consultar de otras maneras ...
// Si no tendríamos un problema de recursividad... Ahora no lo ves, pero seguro que en Acceso a Datos lo ves!!